core/src/tile/tileManager.cpp

```diff
@@ -303,204 +215,244 @@ void TileManager::updateTileSet(TileSet& tileSet) {
-        updateProxyTiles(tileSet, *tile);
+            // Reset tile on potential internal dynamic data set
+            // TODO rename to resetState() to avoid ambiguity
+            tile->resetState();
+        } else {
+            // Clear stale tile data
+            tile.reset();
+        }
     }
 
-    tile->setVisible(true);
-    tileSet.tiles.emplace(_tileID, tile);
+    // Add TileEntry to TileSet
+    auto entry = _tileSet.tiles.emplace(_tileID, tile);
 
-    return cached;
+    if (!tile) {
+        // Add Proxy if corresponding proxy MapTile ready
+        updateProxyTiles(_tileSet, _tileID, entry.first->second);
+    }
+
+    return bool(tile);
 }
 
-void TileManager::removeTile(TileSet& tileSet, std::map<TileID,
-                             std::shared_ptr<Tile>>::iterator& _tileIt,
-                             std::vector<TileID>& _removes) {
+void TileManager::removeTile(TileSet& _tileSet, std::map<TileID, TileEntry>::iterator& _tileIt) {
 
-    const TileID& id = _tileIt->first;
-    auto& tile = _tileIt->second;
+    auto& id = _tileIt->first;
+    auto& entry = _tileIt->second;
 
-    if (tile->hasState(TileState::loading) &&
-        setTileState(*tile, TileState::canceled)) {
-        // 1. Remove from Datasource. Make sure to cancel the network request
-        // associated with this tile.
-        tileSet.source->cancelLoadingTile(id);
-    }
 
-    clearProxyTiles(tileSet, *tile, _removes);
+    if (entry.isLoading()) {
+        entry.cancelTask();
 
-    if (tile->hasState(TileState::ready) && !tile->reloading) {
+        // 1. Remove from Datasource. Make sure to cancel
+        //  the network request associated with this tile.
+        _tileSet.source->cancelLoadingTile(id);
+
+    } else if (entry.isReady()) {
         // Add to cache
-        m_tileCache->put(tileSet.source->id(), tile);
+        m_tileCache->put(_tileSet.source->id(), entry.tile);
     }
 
     // Remove tile from set
-    _tileIt = tileSet.tiles.erase(_tileIt);
+    _tileIt = _tileSet.tiles.erase(_tileIt);
 }
 
-bool TileManager::updateProxyTile(TileSet& tileSet, Tile& _tile, const TileID& _proxy,
-                                  const Tile::ProxyID _proxyID) {
-    auto& tiles = tileSet.tiles;
+bool TileManager::updateProxyTile(TileSet& _tileSet, TileEntry& _tile,
+                                  const TileID& _proxyTileId,
+                                  const ProxyID _proxyId) {
+    auto& tiles = _tileSet.tiles;
 
     // check if the proxy exists in the visible tile set
     {
-        const auto& it = tiles.find(_proxy);
+        const auto& it = tiles.find(_proxyTileId);
         if (it != tiles.end()) {
-            auto& proxyTile = it->second;
-            if (proxyTile && _tile.setProxy(_proxyID)) {
-                proxyTile->incProxyCounter();
-                if (proxyTile->isReady()) {
-                    m_tiles.push_back(proxyTile);
+            // FIXME: this uses indirect proxy tiles
```

Comment:
The general problem is that this logic does not find the nearest available proxy. I try to give an example: Say, tiles from z=0 and z=3 are ready; The view is zooming in from z=0 to z=2 quickly, so that no new tiles get loaded in between. 
At z=1 the direct parent tile is used as proxy(this is ok). But at z=2 the direct parent is used as proxy which is not ready and still references the proxy from z=0, even though the direct proxies from z=3 would be available.